docs/language/informal/toplevel-inference.md

```diff
@@ -0,0 +1,311 @@
+emitting the error for the purposes of emitting subsequent error messages, but
+the behavior of programs for which initializer inference fails is unspecified.
+
+#### Immediately evident expressions
+
+We define a set of expressions with immediately-evident types (called
+immediately-evident expressions, abbreviated IE) as follows.  Immediately
+evident expressions have an inferred type as specified below.  Any expression
+which is not an IE has no inferred type.  For every immediately evident
+expression, we also define the set of variable dependencies upon which its
+inference relies.
+
+
+* null, boolean, numeric, string, and symbol literals have their corresponding
+   type.
+   * No inference dependencies
+* A list literal or a map literal with explicit type arguments has the type
+   `List<T>` or `Map<K, V>` respectively, where `<T>` or `<K, V>` are the
+   provided type arguments.
+   * No inference dependencies
+* A non-empty list literal with no type arguments, and for which all of the
+   elements are IEs, has the type `List<T>` where `T` is the least upper bound
+   of the inferred types of the elements.
+   * If the elements do not all have an inferred type, it is an error
+   * The inference dependencies of the list literal are the collected
+      dependencies of the elements.
+* An non-const empty list literal with no type arguments has the type
+  `List<dynamic>`.
+   * No inference dependencies.
+* A const empty list literal with no type arguments has the type
+  `List<bottom>`.
+   * No inference dependencies.
+* A map literal with no type arguments, and for which all of the keys and
+   values are IEs, has the type `Map<K, V>` where `K` is the least upper bound of
+   the inferred types of the provided keys, and `V` is the least upper bound of
+   the provided values.
+   * If the keys and values do not all have an inferred type, it is an error.
+   * The inference dependencies of the map literal are the collected
+      dependencies of the keys and values.
+* An non-const empty map literal with no type arguments has the type
+  `Map<dynamic, dynamic>`.
+   * No inference dependencies.
+* A const empty map literal with no type arguments has the type `Map<bottom,
+  bottom>`.
+   * No inference dependencies.
+* A type literal has type `Type`
+   * No inference dependencies
+* A function literal with an expression body for which all parameter types are
+   type annotated and for which the return expression is an IE and has an
+   inferred type, has the corresponding function type.
+   * If the return expression has no inferred type it is an error.
+   * The inference dependencies are the inference dependencies of the return
+      expression.
+* An instance creation expression with no omitted type arguments has the
+   obvious type.
+   * No inference dependencies
+* An `as` expression has the cast to type
+   * No inference dependencies
+* A simple or qualified identifier referring to a top level function, static
```

Comment:
I had trouble parsing this section.  My understanding after our conversation today is that:
- References to a top level function, top level variable, top level getter, class static method, class static field, or class static getter are all inferred.
- References to a class instance method are inferred (this is possible because instance inference is always done before static variable and field inference).  Note that we're talking about method tear-offs here.
- References to a class instance field or a class instance getter are not inferred (and result in an error).  This is because we can't support this without losing the ability to compute the dependency graph statically.